### PR TITLE
Incremental save of annotation.

### PR DESCRIPTION
**Pull request by azhavoro**

### CODE REVIEW

File: cvat/apps/engine/static/engine/js/annotationParser.js [hunk 1]
```diff
@@ -77,7 +79,8 @@ class AnnotationParser {
 
         let occluded = +shape.getAttribute('occluded');
         let z_order = shape.getAttribute('z_order') || '0';
-        return [points, occluded, +z_order];
+        let client_id = box.getAttribute('client_id') || '-1'
```
Comment (nmanovic) [inline cvat/apps/engine/static/engine/js/annotationParser.js]: @azhavoro ,

Need to check somehow that client_id is valid. For example, we don't have the same client id for different boxes or all of them are -1. The check should be on client side when we parse annotation and on server side when the user presses down key.

File: cvat/apps/engine/static/engine/js/annotationUI.js [hunk 2]
```diff
@@ -609,11 +609,12 @@ function saveAnnotation(shapeCollectionModel, job) {
...
 
-    let exportedData = shapeCollectionModel.export();
-    let annotationLogs = Logger.getLogs();
+    const exportedData = shapeCollectionModel.export();
+    shapeCollectionModel.reset_state();
+    const annotationLogs = Logger.getLogs();
 
     const data = {
-        annotation: exportedData,
+        annotation: JSON.stringify(exportedData),
```
Comment (nmanovic) [inline cvat/apps/engine/static/engine/js/annotationUI.js]: What is the reason to send it as string?

File: cvat/apps/engine/static/engine/js/shapeCollection.js [hunk 3]
```diff
@@ -221,49 +229,84 @@ class ShapeCollectionModel extends Listener {
...
+            shape_container_target = export_action_container.polygon_paths;
+            break;
+        case 'annotation_polyline':
+            shape_container_target = export_action_container.polylines;
+            break;
+        case 'interpolation_polyline':
+            shape_container_target = export_action_container.polyline_paths;
+        }
+
+        if (!shape_container_target) {
```
Comment (nmanovic) [inline cvat/apps/engine/static/engine/js/shapeCollection.js]: Does switch has 'default'? Does it make sense to move the statement into default?

File: cvat/apps/engine/static/engine/js/shapeCollection.js [hunk 4]
```diff
@@ -221,49 +229,84 @@ class ShapeCollectionModel extends Listener {
...
+            throw Error('Undefined shape type');
+        }
+        return shape_container_target;
+    }
+
+    reset_state() {
+        this._erased = false;
+    }
+
+    export() {
```
Comment (nmanovic) [inline cvat/apps/engine/static/engine/js/shapeCollection.js]: How long does the function works on big tasks like FP/FN?

File: cvat/apps/engine/static/engine/js/annotationUI.js (hunk 2 above)
Comment (azhavoro) [reply cvat/apps/engine/static/engine/js/annotationUI.js]: I think cause of issue was hash function, that used early. I've not changed request format for backward capability and suggest to change it in new MR.

File: cvat/apps/engine/static/engine/js/shapeCollection.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (azhavoro) [reply cvat/apps/engine/static/engine/js/shapeCollection.js]: Thanks, I'll prepare fix for that.

File: cvat/apps/engine/annotation.py [hunk 5]
```diff
@@ -1726,7 +2085,8 @@ def _flip_shape(shape, im_w, im_h):
                                                 "{:.2f}".format(float(p.split(',')[1]))
                                             )) for p in shape.points.split(' '))
                                         )),
-                                        ("occluded", str(int(shape.occluded)))
+                                        ("occluded", str(int(shape.occluded))),
+                                        ("client_id", "{}".format(shape.client_id)),
```
Comment (nmanovic) [inline cvat/apps/engine/annotation.py]: Please use str

File: cvat/apps/engine/annotation.py [hunk 6]
```diff
@@ -1773,7 +2133,8 @@ def _flip_shape(shape, im_w, im_h):
                     for path in path_list:
                         dump_dict = OrderedDict([
                             ("id", str(path_idx)),
-                            ("label", path.label.name)
+                            ("label", path.label.name),
+                            ("client_id", "{}".format(path.client_id)),
```
Comment (nmanovic) [inline cvat/apps/engine/annotation.py]: Please use str

**Comment by bsekachev** [comment]
@azhavoro 
I found bug in this patch. I tried next scenario:
1) I opened a annotation task.
2) I remove all annotation by click "Remove annotation" button
3) I created any bounding box
4) I saved it and got 400 error.
5) I saved it again and it was successful.
6) I updated page and I saw old annotation together with new annotation.

**Comment by bsekachev** [comment]
@azhavoro 
Also after client parsed an annotation file without client_id field, we have situation when all objects have index -1. I think it isn't right.

**Comment by bsekachev** [comment]
@azhavoro 
Other bug in a scenario:
1) I opened task (it was a interpolation task, but I think it doesn't matter)
2) I uploaded annotation file without client_id fields. 
3) I saved job. It was successful. 
4) I moved any box. 
5) I saved it again and I got 400 error. 
6) Each next save also returned 400 code.

**Comment by bsekachev** [comment]
@azhavoro 
ReID doesn't work with this patch. I checked it.

**Pull request merged**